docs/modules/cluster-performance/pages/thread-per-core-tpc.adoc

```diff
@@ -0,0 +1,97 @@
+= Thread-Per-Core (TPC)
+:description: Find out more about Thread-Per-Core (TPC) and how to enable this feature on clients and cluster members.
+:page-enterprise: true
+
+Enable Thread-Per-Core (TPC) on both clients and cluster members to improve system performance, especially on machines with either fewer or more CPU cores than the recommended hardware listed on the xref:ROOT:production-checklist.adoc[].  
+
+[[tpc-what]]
+== What is Thread-Per-Core (TPC)?
+
+The threading model is responsible for networking, storage and compute, which means it has a big impact on system performance. With Staged Event-Driven Architecture (SEDA) you have dedicated threads that do one particular task (for example, reading from a socket or evaluating a map.get) and thread pools that manage these threads.  
+
+SEDA can lead to throughput and latency issues on machines with limited cores (fewer than 8), as the frequent context switching and state sharing can add significant overhead. Conversely, on machines with a very high number of CPUs (more than 32), a bottleneck can be created by the limited number of I/O threads vs. operation threads. Typically, smaller servers are more readily available through virtualization and for cloud-specific workloads, which can limit system performance in a SEDA environment. 
```

Comment:
The bottleneck on more than 32 stated here should be reflected in the Hardware Recommendations of the Production Checklist topic.

And, you know I'm going to complain about vs. - how about compared to? Or, "a bottleneck can occur due to the difference in the available threads used for IO and operations"?

What do you mean by "readily available"? They remain available for use, they are easier to setup/manage/use in virtual and cloud environments, or they are more commonly used in these environments?

Reply:
> The bottleneck on more than 32 stated here should be reflected in the Hardware Recommendations of the Production Checklist topic.

Possibly...I was trying to keep the TPC changes controlled as I don't know if we're ready to change our overall recommendations (slightly more ambitious scope). We should have a link to TPC from checklist topic anyway.

What do you think @pveentjer ?

I think the 'readily available' comment refers to cost/availability considerations when using cloud i.e. you pay more for higher core machines so bang per buck does lead to less cores.

(vs. is listed in the MS guide and you know I'm going to highlight the inconsistency vs. i.e. and e.g.)

Reply:
There is no simple answer. If the workload is CPU intensive like query processing, TPC will not make any difference. So in that case, with TPC disabled, going for a heavy machine still makes sense.

But if the interactions are simple point interactions (not long-running tasks) then yes, the higher the CPU count, you quickly run into diminishing returns on heavier machines with classic.

Also it depends on the workload. In 5.4 there are only TPC connections between clients/members. And this will speed up readonly operations because they are not bottlenecked by the member/member connections.

If it will be decided to include member/member TPC connections, also update heavy workloads, dumb clients, CP can speed up because they rely on the member/member connection which very quickly becomes a bottleneck.